### PR TITLE
【OSCP】 在 SecretFlow 中添加基于torch后端的fed_pa​​c策略

### PR DESCRIPTION
**Pull request by Dorothy21650**
**Type of change**

- [ ] Add new papers (Please tell us why you think this paper is awesome!)
- [ ] Fix the category of an existing paper/papers (Please tell us the reasons)
- [ x] Add a new tool/primitive/application with a new markdown page (Thank you! Also, please tell us more about this awesome thing!)

**Description**

> Fixed https://github.com/secretflow/secretflow/issues/1280

### CODE REVIEW

**Comment by Dorothy21650** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by wyanbing** [comment]
I have read the CLA Document and I hereby sign the CLA

File: secretflow/ml/nn/core/torch/module.py [hunk 1]
```diff
@@ -244,6 +245,92 @@ def update_metrics(self, y_pred, y_true):
                 m.update(y_pred, y_true.int())
 
 
+class CifarCNN(BaseModule):
```
Comment (ian-huu) [inline secretflow/ml/nn/core/torch/module.py:248]: 模型定义到 application 里，不要放在这里

File: secretflow/ml/nn/fl/backend/torch/fl_base.py [hunk 2]
```diff
@@ -420,3 +433,303 @@ def load_model(self, model_path: str, **kwargs):
         if self.use_gpu:
             self.model.to(self.exe_device)
         return checkpoint['epoch']
+
+
+from torch import nn
+from copy import deepcopy
+
+
+class FedPACTorchModel(BaseTorchModel):
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/fl/backend/torch/fl_base.py:442]: 这个类可以合到 strategy/fed_pac.py FedPAC

File: secretflow/ml/nn/fl/backend/torch/strategy/fed_pac.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        model.zero_grad()
+        global_protos = self.global_protos
+        self.last_model = deepcopy(model)
+        # get local prototypes before training, dict:={label: list of sample features}
+        local_protos1 = self.get_local_protos(step, train_steps)
+        # Set optimizer for the local updates, default sgd
+        lr = kwargs.get("lr", 0.01)
+
+        epoch_classifier = 1
+        optimizer = torch.optim.SGD(
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/fl/backend/torch/strategy/fed_pac.py:67]: optimizer loss 之类的可以写到 TorchModel 里吗

File: secretflow/ml/nn/fl/backend/torch/strategy/fed_pac.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+            else:
+                param.requires_grad = False
+        lr_g = 0.1
+        optimizer = torch.optim.SGD(
+            filter(lambda p: p.requires_grad, model.parameters()),
+            lr=lr_g,
+            momentum=0.5,
+            weight_decay=0.0005,
+        )
+        for ep in range(epoch_classifier):
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/fl/backend/torch/strategy/fed_pac.py:82]: 每个 step 训练一个 epoch 吗

File: secretflow/security/aggregation/fedpac_aggregator.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+# from copy import copy, deepcopy
+import math
+
+
+from secretflow.device import PYU, DeviceObject, PYUObject
+from secretflow.security.aggregation.aggregator import Aggregator
+
+
+class FedPACAggregator(Aggregator):
```
Comment (zhaocaibei123) [inline secretflow/security/aggregation/fedpac_aggregator.py:31]: average sum 逻辑没变的话可以继承 PlainAggregator

**Review by zhaocaibei123 — CHANGES_REQUESTED** [review]
加个单测，参考 tests/ml/nn/fl/test_fl_model_torch.py

File: secretflow/ml/nn/fl/backend/torch/fl_base.py [hunk 6]
```diff
@@ -420,3 +433,303 @@ def load_model(self, model_path: str, **kwargs):
...
+                if s_w is not None and len(s_w.shape) > 0:
+                    x.append(s_w)
+
+            self.train_set, self.eval_set, step_per_epoch = dataset_builder(
+                x, stage=stage
+            )
+
+        if stage != "train" and stage != "eval":
+            raise Exception(f"Illegal argument stage={stage}")
+        return step_per_epoch
```
[the remark this answers is not in the document]
Comment (Dorothy21650) [reply secretflow/ml/nn/fl/backend/torch/fl_base.py]: 已经修改（删除掉了，用的flbase中原有的做法），之前是为了跟源码里面对齐（源码中有在local_training中进行local_test），迁移过来之后，这个逻辑也删除了。

File: secretflow/ml/nn/fl/backend/torch/fl_base.py [hunk 7]
```diff
@@ -49,6 +51,7 @@ def __init__(
         self.train_set = None
         self.eval_set = None
         self.skip_bn = skip_bn
+        # self.dataset_size = torch.tensor(len(self.train_set.dataset)).to(self.exe_device)
```
[the remark this answers is not in the document]
Comment (Dorothy21650) [reply secretflow/ml/nn/fl/backend/torch/fl_base.py:54]: Done

File: secretflow/ml/nn/core/torch/module.py [hunk 8]
```diff
@@ -244,6 +245,92 @@ def update_metrics(self, y_pred, y_true):
...
+    def forward(self, x):
+        x = self.pool(F.leaky_relu(self.conv1(x)))
+        x = self.pool(F.leaky_relu(self.conv2(x)))
+        x = self.pool(F.leaky_relu(self.conv3(x)))
+        x = x.view(-1, self.num_flat_features(x))
+        x = F.leaky_relu(self.fc1(x))
+        y = self.fc2(x)
+        return x, y
+
+    def feature2logit(self, x):
```
[the remark this answers is not in the document]
Comment (Dorothy21650) [reply secretflow/ml/nn/core/torch/module.py]: Done.

File: secretflow/ml/nn/core/torch/module.py (hunk 1 above)
Comment (Dorothy21650) [reply secretflow/ml/nn/core/torch/module.py]: Done.

File: secretflow/ml/nn/fl/backend/torch/fl_base.py [hunk 9]
```diff
@@ -420,3 +433,303 @@ def load_model(self, model_path: str, **kwargs):
...
+        builder_base: BuilderType,
+        random_seed: int = None,
+        skip_bn: bool = False,
+        **kwargs,
+    ):
+        super().__init__(builder_base, random_seed=random_seed, **kwargs)
+        self.num_classes = kwargs.get("num_classes", 10)
+        self.criterion = nn.CrossEntropyLoss()
+        self.local_model = self.model
+        self.last_model = deepcopy(self.model)
```
[the remark this answers is not in the document]
Comment (Dorothy21650) [reply secretflow/ml/nn/fl/backend/torch/fl_base.py]: 这里是照搬源码的，源码是在本地训练开始之前保存的模型的深度副本，可能是考虑到了回滚，但是源码中也没有明确使用 self.last_model，现在已经删除掉这行了。

File: secretflow/ml/nn/fl/backend/torch/fl_base.py (hunk 2 above)
Comment (Dorothy21650) [reply secretflow/ml/nn/fl/backend/torch/fl_base.py]: Done.

File: secretflow/ml/nn/fl/backend/torch/fl_base.py [hunk 10]
```diff
@@ -216,9 +219,9 @@ def get_rows_count(self, filename):
 
     def get_weights(self, return_numpy=True):
         if self.skip_bn:
-            return self.model.get_weights_not_bn(return_numpy=return_numpy)
+            return self.model.get_weights_not_bn(return_numpy=True)
```
Comment (zhaocaibei123) [inline secretflow/ml/nn/fl/backend/torch/fl_base.py:222]: 这里 return_numpy 为什么写死？

**Comment by Candicepan** [comment]
续一下
